druid/src/text/editor.rs

```diff
@@ -0,0 +1,311 @@
+                let to_delete = if self.selection.is_caret() {
+                    movement(mvmt, self.selection, data, true)
+                } else {
+                    self.selection
+                };
+                data.edit(to_delete.range(), "");
+                self.selection = Selection::caret(to_delete.min());
+            }
+            EditAction::Move(mvmt) => self.selection = movement(mvmt, self.selection, data, false),
+            EditAction::ModifySelection(mvmt) => {
+                self.selection = movement(mvmt, self.selection, data, true)
+            }
+            EditAction::Click(action) => {
+                if action.mods.shift() {
+                    self.selection.end = action.column;
+                } else {
+                    self.selection = Selection::caret(action.column);
+                }
+            }
+            EditAction::Drag(action) => self.selection.end = action.column,
+            EditAction::SelectAll => self.selection = Selection::new(0, data.len()),
+        }
+    }
+
+    /// Draw this editor at the provided point.
+    pub fn draw(&self, ctx: &mut PaintCtx, point: impl Into<Point>) {
+        self.layout.draw(ctx, point)
+    }
+
+    /// Returns `true` if the data passed here has been changed externally,
+    /// which means things like our selection state may be out of sync.
+    ///
+    /// This would only happen in the unlikely case that somebody else has mutated
+    /// the data before us while handling an event; if this is the case we ignore
+    /// the event, and our data will be updated in `update`.
+    fn data_is_stale(&self, data: &T) -> bool {
+        self.layout.text().map(|t| !t.same(data)).unwrap_or(true)
+    }
+
+    fn insert(&mut self, data: &mut T, text: &str) {
+        // if we aren't multiline, we insert only up to the first newline
+        let text = if self.multiline {
+            text
+        } else {
+            text.split('\n').next().unwrap_or("")
+        };
+        let sel = self.selection.range();
+        data.edit(sel, text);
+        self.selection = Selection::caret(self.selection.min() + text.len());
+    }
+
+    /// Delete to previous grapheme if in caret mode.
+    /// Otherwise just delete everything inside the selection.
+    fn delete_backward(&mut self, data: &mut T) {
+        let cursor_pos = if self.selection.is_caret() {
+            let del_end = self.selection.end;
+            let del_start = offset_for_delete_backwards(&self.selection, data);
+            data.edit(del_start..del_end, "");
+            del_start
+        } else {
+            data.edit(self.selection.range(), "");
+            self.selection.min()
+        };
+
+        self.selection = Selection::caret(cursor_pos);
+    }
+
+    fn delete_forward(&mut self, data: &mut T) {
+        let to_delete = if self.selection.is_caret() {
+            movement(Movement::Right, self.selection, data, false)
```

Comment:
Delete button doesn't do anything when text isn't selected.

The `self.selection` starts and ends at the same index, e.g. `Selection { start: 16, end: 16 }`. I think you might need something like `offset_for_delete_forward`.

Reply:
oops, the `false` should have been `true`.